// stryker-javascript-mutator/src/mutators/StringLiteralMutator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var types = require("@babel/types");
var StringLiteralMutator = /** @class */ (function () {
    function StringLiteralMutator() {
        this.name = 'StringLiteral';
    }
    StringLiteralMutator.prototype.mutate = function (node, copy) {
        var nodes = [];
        if (types.isTemplateLiteral(node)) {
            var mutatedNode = {
                end: node.end,
                innerComments: node.innerComments,
                leadingComments: node.leadingComments,
                loc: node.loc,
                start: node.start,
                trailingComments: node.trailingComments,
                type: 'StringLiteral',
                value: ''
            };
            if (node.quasis.length === 1 && node.quasis[0].value.raw.length === 0) {
                mutatedNode.value = 'Stryker was here!';
            }
            nodes.push(mutatedNode);
        }
        else if ((!node.parent || (!types.isImportDeclaration(node.parent) && !types.isJSXAttribute(node.parent)))
            && types.isStringLiteral(node)) {
            var mutatedNode = copy(node);
            mutatedNode.value = mutatedNode.value.length === 0 ? 'Stryker was here!' : '';
            nodes.push(mutatedNode);
        }
        return nodes;
    };
    return StringLiteralMutator;
}());
exports.default = StringLiteralMutator;
