// stryker/src/initializer/StrykerConfigWriter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var _ = require("lodash");
var util_1 = require("@stryker-mutator/util");
var logging_1 = require("stryker-api/logging");
var prettier_1 = require("prettier");
var STRYKER_CONFIG_FILE = 'stryker.conf.js';
var StrykerConfigWriter = /** @class */ (function () {
    function StrykerConfigWriter(out) {
        this.out = out;
        this.log = logging_1.getLogger(StrykerConfigWriter.name);
    }
    StrykerConfigWriter.prototype.guardForExistingConfig = function () {
        if (util_1.fsAsPromised.existsSync(STRYKER_CONFIG_FILE)) {
            var msg = 'Stryker config file "stryker.conf.js" already exists in the current directory. Please remove it and try again.';
            this.log.error(msg);
            throw new Error(msg);
        }
    };
    /**
     * Create stryker.conf.js based on the chosen framework and test runner
     * @function
     */
    StrykerConfigWriter.prototype.write = function (selectedTestRunner, selectedTestFramework, selectedMutator, selectedTranspilers, selectedReporters, selectedPackageManager, additionalPiecesOfConfig) {
        var configObject = {
            mutator: selectedMutator ? selectedMutator.name : '',
            packageManager: selectedPackageManager.name,
            reporters: selectedReporters.map(function (rep) { return rep.name; }),
            testRunner: selectedTestRunner ? selectedTestRunner.name : '',
            transpilers: selectedTranspilers ? selectedTranspilers.map(function (t) { return t.name; }) : []
        };
        this.configureTestFramework(configObject, selectedTestFramework);
        _.assign.apply(_, [configObject].concat(additionalPiecesOfConfig));
        return this.writeStrykerConfig(configObject);
    };
    StrykerConfigWriter.prototype.configureTestFramework = function (configObject, selectedTestFramework) {
        if (selectedTestFramework) {
            configObject.testFramework = selectedTestFramework.name;
            configObject.coverageAnalysis = 'perTest';
        }
        else {
            configObject.coverageAnalysis = 'all';
        }
    };
    StrykerConfigWriter.prototype.writeStrykerConfig = function (configObject) {
        this.out('Writing stryker.conf.js...');
        return util_1.fsAsPromised.writeFile(STRYKER_CONFIG_FILE, this.wrapInModule(configObject));
    };
    StrykerConfigWriter.prototype.wrapInModule = function (configObject) {
        return prettier_1.format("\n      module.exports = function(config){\n        config.set(\n          " + JSON.stringify(configObject, null, 2) + "\n        );\n      }", { parser: 'babylon' });
    };
    return StrykerConfigWriter;
}());
exports.default = StrykerConfigWriter;
